### ISSUE TITLE
tsc was not available or found in your node_modules or $PATH. Please run install and try again. 

### ISSUE THREAD

**Issue by Armadillidiid** [dmmulroy/tsc.nvim#28]
I get this error in my typescript vite project. It was working before for a long time until after I updated yesterday. Here is my package.json:

  "packages": {
    "": {
      "dependencies": {
        "@headlessui/react": "^1.7.14",
        "@heroicons/react": "^2.0.17",
        "@tailwindcss/forms": "^0.5.3",
        "@tanstack/react-query": "^4.29.13",
        "@tanstack/react-query-devtools": "^4.29.14",
        "@tanstack/react-table": "^8.9.1",
        "@types/crypto-js": "^4.1.1",
        "@types/react-csv": "^1.1.3",
        "axios": "^1.4.0",
        "class-variance-authority": "^0.6.0",
        "clsx": "^1.2.1",
        "color": "^4.2.3",
        "crypto-js": "^4.1.1",
        "date-fns": "^2.30.0",
        "framer-motion": "^10.12.16",
        "jwt-decode": "^3.1.2",
        "react": "^18.2.0",
        "react-csv": "^2.2.2",
        "react-day-picker": "^8.7.1",
        "react-dom": "^18.2.0",
        "react-hook-form": "^7.43.9",
        "react-icons": "^4.8.0",
        "react-router-dom": "^6.11.2",
        "react-toastify": "^9.1.3",
        "recharts": "^2.5.0",
        "tailwind-merge": "^1.12.0",
        "vite-plugin-svgr": "^3.2.0",
        "zod": "^3.21.4"
      },
      "devDependencies": {
        "@faker-js/faker": "^8.0.0",
        "@types/react": "^18.0.28",
        "@types/react-dom": "^18.0.11",
        "@typescript-eslint/eslint-plugin": "^5.57.1",
        "@typescript-eslint/parser": "^5.57.1",
        "@vitejs/plugin-react-swc": "^3.0.0",
        "autoprefixer": "^10.4.14",
        "eslint": "^8.38.0",
        "eslint-plugin-react-hooks": "^4.6.0",
        "eslint-plugin-react-refresh": "^0.3.4",
        "postcss": "^8.4.23",
        "prettier": "^2.8.8",
        "prettier-plugin-tailwindcss": "^0.3.0",
        "tailwindcss": "^3.3.2",
        "typescript": "^5.0.2",
        "vite": "^4.3.2"
      }
    }

**Comment by dmmulroy** [comment]
Can you share your tsconfig(s) and the project structure? In the mean time you can try setting the `bin_path` configuration option.

**Comment by Armadillidiid** [comment]
It's a company repository, so unfortunately, I am unable to provide detailed information about the project structure. However, I can tell you that the tsconfig file is generally the default configuration generated by Vite when initializing a new project.

Regarding your question about setting the bin_path configuration, could you please provide more context or specify how I could do that?

**Comment by dmmulroy** [comment]
There is a `bin_path` config option that you can set to the path to your `tsc` binary. See #26 for more details.

**Comment by Armadillidiid** [comment]
I get the following error with the default configuration in the docs. Do I have to manually implement the `utils` function?

####  Error notify.error lazy.nvim Failed to run `config` for tsc.nvim
 
`
/home/xxx/.config/LazyVim/lua/plugins/tsc-nvim.lua:7: attempt to index global 'utils' (a nil value)`

### Config:
```typescript
return {
  "dmmulroy/tsc.nvim",
  config = function()
    require("tsc").setup({
      auto_open_qflist = true,
      auto_close_qflist = false,
      bin_path = utils.find_tsc_bin(),
      enable_progress_notifications = true,
      flags = {
        noEmit = true,
        project = function()
          return utils.find_nearest_tsconfig()
        end,
      },
      hide_progress_notifications_from_history = true,
      spinner = { "⣾", "⣽", "⣻", "⢿", "⡿", "⣟", "⣯", "⣷" },
    })
  end,
}

```

**Comment by dmmulroy** [comment]
I also just installed a fresh project using `npm create vite@latest my-vite-app -- --template vanilla-typescript` and everything worked, so you may need to add some logging to the plugin or call `utils.find_tsc_bin()` from the status line cmdline (e.g. `:lua print(vim.inspect(require('tsc.utils').find_tsc_bin()))`)

**Comment by Armadillidiid** [comment]
> 

No, invoking `which tsc` returns `~/.local/bin/tsc`, which is a symlink to `/lib/node_modules/typescript/bin/tsc` . Therefore, I opted to use that instead. In any case, I managed to get it working by utilizing the utility function you provided. I had ChatGPT to make any corrections if needed and it worked. Specifying the `bin_path` never worked, nor did requiring the `tsc.utils`.

Here's the working configuration:

```typescript
local utils = {}

utils.find_tsc_bin = function()
  local node_modules_tsc_binary = vim.fn.findfile("node_modules/.bin/tsc", ".;")

  if node_modules_tsc_binary ~= "" then
    return node_modules_tsc_binary
  end

  return "tsc"
end

return {
  "dmmulroy/tsc.nvim",
  config = function()
    require("tsc").setup({
      bin_path = utils.find_tsc_bin(),
    })
  end
}

**Comment by dmmulroy** [comment]
Ah, I wonder if the problem is the symlink. I'm betting that `utils.is_executable(tsc)` doesn't follow the symlink and returns `false`.

**Comment by dmmulroy** [comment]
That's really bizarre. You _should_ at least be able to simplify that to just:

```lua
return {
  "dmmulroy/tsc.nvim",
  config = function()
    require("tsc").setup({
      bin_path = require('tsc.utils').find_tsc_bin(),
    })
  end
}
```

I wonder if `vim.fn.executable` is doing something unexpected

> executable({expr})					*executable()*
		This function checks if an executable with the name {expr}
		exists.  {expr} must be the name of the program without any
		arguments.
		executable() uses the value of $PATH and/or the normal
		searchpath for programs.		*PATHEXT*
		On MS-Windows the ".exe", ".bat", etc. can optionally be
		included.  Then the extensions in $PATHEXT are tried.  Thus if
		"foo.exe" does not exist, "foo.exe.bat" can be found.  If
		$PATHEXT is not set then ".exe;.com;.bat;.cmd" is used.  A dot
		by itself can be used in $PATHEXT to try using the name
		without an extension.  When 'shell' looks like a Unix shell,
		then the name is also tried without adding an extension.
		On MS-Windows it only checks if the file exists and is not a
		directory, not if it's really executable.
		On Windows an executable in the same directory as Vim is
		always found (it is added to $PATH at |startup|).
		The result is a Number:
			1	exists
			0	does not exist
			-1	not implemented on this system
		|exepath()| can be used to get the full path of an executable.

**Comment by evanrichards** [comment]
I was also hitting this issue. For whatever reason running  `:echo vim.fn.findfile("node_modules/.bin/tsc", ".;")` gave me nothing, but running `:echo vim.fn.findfile("node_modules/.bin/tsc")` spit out the correct tsc path. I updated my config setup to look like this: 
```lua
require("tsc").setup({
	auto_open_qflist = true,
	auto_close_qflist = false,
	bin_path = vim.fn.findfile("node_modules/.bin/tsc"),
	enable_progress_notifications = true,
	flags = {
		noEmit = true,
		project = function()
			return vim.fn.findfile("tsconfig.json")
		end,
	},
	hide_progress_notifications_from_history = true,
	spinner = { "⣾", "⣽", "⣻", "⢿", "⡿", "⣟", "⣯", "⣷" },
})
```

note that i hit the same issue in the project file flag config.


https://github.com/evanrichards/dotfiles/commit/08fe91648f6fdeff711e3bfc149eab035cf0dc0a

**Comment by dmmulroy** [comment]
@evanrichards do you think you could replicate the issue in a small repo and post it?

**Comment by sami616** [comment]
Very strange this also seems to be an issue here, im working in a pnpm monorepo with `build = true` set.

**Comment by dmmulroy** [comment]
@sami616 Could you create a small pnpm monorepo where this is reproducible? I've never used pnpm or it's monorepo capabilities before.

**Comment by dmmulroy** [comment]
@sami616 could you try using this config
```lua
require("tsc").setup({
	flags = {
		noEmit = false,
		build = true,
		project = false,
	},
})
```

**Comment by GitMurf** [comment]
@dmmulroy fyi I just started using your awesome plugin and I ran into similar issues. I also use `pnpm` as someone above mentioned but I don't think that should be a problem. I think potentially the problem is due to windows and likely path issues (backslashes vs forward, escaping, spaces in paths etc.). This is the most common issue I have ran into with many plugins in the neovim community. Curious if you have had a chance to test on windows?

That all being said, my solution was that I could use the defaults as provided in the readme except I had to change it to use `bin_path = "tsc"`. All is working nicely after making this change.

**Comment by dmmulroy** [comment]
@GitMurf thanks for the details! Pnpm and workspaces have been a thorn in my side for this plugin haha. Is there any chance you could create a small repo where the bug is reproducible?

**Comment by GitMurf** [comment]
> @GitMurf thanks for the details! Pnpm and workspaces have been a thorn in my side for this plugin haha. Is there any chance you could create a small repo where the bug is reproducible?

Yes but may take me a week or so. Swamped right now as I'm sure you understand ;)

Again, awesome work on this plugin! Absolute game changer for me for typescript development :)

**Comment by HadiModarres** [comment]
I cannot run `TSC` as it complains tsc was not found. 
Why can't we just used `npx tsc` to find where tsc is automatically, instead of providing it manually through bin_path?

**Comment by nikolaywithpara** [comment]
Maybe will help somebody, you might open the project with `nvim .` and because of that for some reason plugin doesn't find tsc bin. If I open project with just `nvim` everything works fine

**Comment by Kifial** [comment]
Maybe will help somebody, you might open the project with nvim . and because of that for some reason plugin doesn't find tsc bin. If I open project with just nvim everything works fine
